### PR TITLE
feat(bot.py): fix type issues

### PR DESCRIPTION
**Pull request by EmreTech**
## Summary

This PR fixes the (many) type issues in `bot.py` and a few type issues in `commands/core.py`.
This PR also makes the code more efficient and readable with issues that aren't entirely type related.

## Information

- [x] This PR fixes an issue.
- [ ] This PR adds something new (e.g. new method or parameters).
- [x] This PR is a breaking change (e.g. methods or parameters removed/renamed).
- [ ] This PR is **not** a code change (e.g. documentation, README, typehinting, examples, ...).

## Checklist

- [x] I have searched the open pull requests for duplicates.
- [ ] If code changes were made then they have been tested.
  - [x] I have updated the documentation to reflect the changes.
- [x] If `type: ignore` comments were used, a comment is also left explaining why.

### CODE REVIEW

**Review by searchify7 — DISMISSED** [review]

**Review by BobDotCom — CHANGES_REQUESTED** [review]

**Comment by BobDotCom** [comment]
Please resolve merge conflicts

File: discord/bot.py [hunk 1]
```diff
@@ -379,7 +382,7 @@ async def register_command(
             The command that was registered
         """
         # TODO: Write this
-        raise RuntimeError("This function has not been implemented yet")
+        raise NotImplementedError
```
[the remark this answers is not in the document]
Comment (EmreTech) [reply discord/bot.py:385]: But it makes a lot more sense if you think about it
The name NotImplementedError implies that the function/class hasn't been implemented yet.

**Comment by BobDotCom** [comment]
Is this finished?

**Comment by EmreTech** [comment]
> Is this finished?

Nope, hence why it's still marked as a draft

**Comment by EmreTech** [comment]
This is now ready for review.

**Review by Lulalaby — APPROVED** [review]
GJ

**Review by Dorukyum — APPROVED** [review]
LGTM, thank you

**Pull request merged**